### ISSUE TITLE
"Not found" screen needs to be more helpful

### ISSUE THREAD

**Issue by allisonkarlitskaya** [cockpit-project/cockpit#16775]
I've had two experiences in the past couple of weeks (one on IRC, one filed as flathub/org.cockpit_project.CockpitClient#2) where I've witnessed users getting confused about the "Not Found" screen due to `cockpit-system` not being installed.

We need to detect if the user is trying to access a path that requires this package to be installed (like `/`) and throw a better error message . During stand-up yesterday, @garrett also mentioned the possibility to offer one-click installation of the required package, but this will require some technical feasibility investigation due to the limited abilities of our error pages.

**Comment by martinpitt** [comment]
Thanks for pointing this out! Indeed, let's fix this.

One-click installation is too complex IMHO -- that requires lots of JS code, interfacing packagekit and the like, which needs to be in a webpack. We could ship that page in cockpit-bridge, but I believe that's overkill. If the error message would say "You need to install the cockpit-system package", this is good enough for almost everyone. The point is not to teach people how to run GNOME software or `apt`/`dnf` on the command line, but to explain what the root cause is.

If we just show that message for path == /, this should work for the majority of cases. One error report had it for http://localhost:9090/cockpit/@localhost/podman/index.html# , but that URL was not specified by a human, but some desktop applet (podman CRC) -- but that demo was just missing the installation of cockpit-podman. Again it's too complex to map out every possible path to its package on a static error page, but I dont' think we have to.

**Comment by garrett** [comment]
What do you think of using an [empty state](https://www.patternfly.org/v4/components/empty-state/) that looks like this?

![image](https://user-images.githubusercontent.com/10246/148079150-fce8a45f-1515-4fd4-a064-ec74bef2651a.png)

Text:

> ### Missing system package
> Please install the `cockpit-system` package with your package manager.

If we could use the one-click installation, we'd have a button to install it as the default action. Since we can't, then we can use some text like in the screenshot.

**Comment by garrett** [comment]
It would be lovely if we could actually give the command based on the OS, but that gets into having all the dnf/yum/apt variants and:

> The point is not to teach people [...] the command line

**Comment by allisonkarlitskaya** [comment]
> Please install the `cockpit-system` package with your package manager.

I wonder if we need to somehow emphasize that this needs to be done on the server you're trying to connect to.  This might otherwise be confusing for Cockpit Client users, for example.

I can't think of any elegant way to phrase that, though.

**Comment by allisonkarlitskaya** [comment]
Also, a small technical note:

this page is generated by templating, based on `fail.html`, which contains this:

```
    <title>@@message@@</title>
```

and this:

```
        <h1>@@message@@</h1>
```

and nothing else.

We can add an "extra text" field if we want, but it's non-trivial.

**Comment by allisonkarlitskaya** [comment]
Also note: this would be a particularly awful time to include a new separate failure page.  I just ripped out GResource support based on the fact that we only ever had a single file in there and it was easier to deal with that manually.  I'd prefer not to add a second one. :)

**Comment by garrett** [comment]
>     Please install the `cockpit-system` package on your server.

Having both package manager and server is weird no matter which way I combine them. I'm a bit sick though, so I could be wrong. :wink:

**Comment by garrett** [comment]
This is standard PF4. Here's an error in the design guidelines: https://www.patternfly.org/v4/components/empty-state/design-guidelines#back-end-failure

If our failure page doesn't support this, then we need to fix it, I guess.

**Comment by allisonkarlitskaya** [comment]
The first go at a fix in #16782 will hopefully make today's release.

**Comment by martinpitt** [comment]
IMHO #16782 fixed the "90%" case, and going further is too complicated, see [comment above](https://github.com/cockpit-project/cockpit/issues/16775#issuecomment-1004766293).
